File: base_agent.py
#!/usr/bin/env python3
"""
Base Agent for Formul8 Platform

Common base class and utilities for all Formul8 agents.
"""

import json
from typing import Dict, Any, Optional
import openai
from langchain_openai import ChatOpenAI
from langchain.schema import HumanMessage, SystemMessage

class AgentResponse:
    """Standard response format for all agents."""
    
    def __init__(self, agent: str, response: str, confidence: float, 
                 sources: Optional[list] = None, metadata: Optional[Dict[str, Any]] = None,
                 requires_human_verification: bool = False):
        self.agent = agent
        self.response = response
        self.confidence = confidence
        self.sources = sources or []
        self.metadata = metadata or {}
        self.requires_human_verification = requires_human_verification
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary format."""
        return {
            "agent": self.agent,
            "response": self.response,
            "confidence": self.confidence,
            "sources": self.sources,
            "metadata": self.metadata,
            "requiresHumanVerification": self.requires_human_verification
        }

class BaseAgent:
    """Base class for all Formul8 agents."""
    
    def __init__(self, agent_type: str, system_prompt: str, api_key: Optional[str] = None):
        """Initialize the base agent."""
        self.agent_type = agent_type
        self.system_prompt = system_prompt
        
        if api_key:
            openai.api_key = api_key
            self.llm = ChatOpenAI(api_key=api_key, model="gpt-4o", temperature=0.3)
        else:
            self.llm = None
    
    def process_query(self, query: str, context: Optional[Dict[str, Any]] = None) -> AgentResponse:
        """Process a query and return a standardized response."""
        try:
            if self.llm is None:
                return AgentResponse(
                    agent=self.agent_type,
                    response="OpenAI API key not configured. Please set OPENAI_API_KEY environment variable.",
                    confidence=0,
                    metadata={"error": "API key not configured"}
                )
            
            # Prepare the prompt
            messages = [
                SystemMessage(content=self.system_prompt),
                HumanMessage(content=query)
            ]
            
            # Get response from OpenAI
            response = self.llm.invoke(messages)
            
            # Parse and structure the response
            result = AgentResponse(
                agent=self.agent_type,
                response=response.content,
                confidence=self._calculate_confidence(response.content),
                sources=self._extract_sources(response.content),
                metadata={
                    "query": query,
                    "context": context,
                    "model": "gpt-4o"
                }
            )
            
            return result
            
        except Exception as e:
            return AgentResponse(
                agent=self.agent_type,
                response=f"Error processing query: {str(e)}",
                confidence=0,
                metadata={"error": str(e)}
            )
    
    def _calculate_confidence(self, response: str) -> float:
        """Calculate confidence score for the response."""
        if not response or len(response) < 10:
            return 0.0
        
        # Simple heuristic: longer, more detailed responses get higher confidence
        confidence = min(95.0, max(50.0, len(response) / 10))
        
        return min(95.0, confidence)
    
    def _extract_sources(self, response: str) -> list:
        """Extract potential sources from the response."""
        sources = []
        
        # Look for common source patterns
        source_patterns = ['http', 'www', 'doi:', 'PMID:', 'PMC']
        for pattern in source_patterns:
            if pattern.lower() in response.lower():
                sources.append(pattern)
        
        return list(set(sources))
    
    def run_baseline_tests(self, test_queries: list) -> Dict[str, Any]:
        """Run baseline tests to evaluate agent performance."""
        results = []
        for query in test_queries:
            result = self.process_query(query)
            results.append({
                "query": query,
                "confidence": result.confidence,
                "response_length": len(result.response)
            })
        
        avg_confidence = sum(r["confidence"] for r in results) / len(results)
        
        return {
            "total_tests": len(test_queries),
            "average_confidence": avg_confidence,
            "results": results
        } 

File: docker-build-context/server/agents/base/base-agent.ts
import OpenAI from "openai";

const openai = process.env.OPENAI_API_KEY ? new OpenAI({ apiKey: process.env.OPENAI_API_KEY }) : null;

export interface AgentResponse {
  agent: string;
  response: string;
  confidence: number;
  sources?: string[];
  metadata?: Record<string, any>;
  requiresHumanVerification?: boolean;
}

export interface VerificationResult {
  consensusReached: boolean;
  finalConfidence: number;
  discrepancies?: string[];
  recommendation: string;
}

export abstract class BaseAgent {
  protected agentType: string;
  protected systemPrompt: string;

  constructor(agentType: string, systemPrompt: string) {
    this.agentType = agentType;
    this.systemPrompt = systemPrompt;
  }

  abstract processQuery(query: string, context?: Record<string, any>): Promise<AgentResponse>;

  protected async callOpenAI(prompt: string, useJson = true): Promise<any> {
    if (!openai) {
      throw new Error("OpenAI API key not configured");
    }

    try {
      const messages = [
        { role: "system" as const, content: this.systemPrompt },
        { role: "user" as const, content: prompt }
      ];

      const response = await openai.chat.completions.create({
        model: "gpt-4o", // newest OpenAI model is "gpt-4o" which was released May 13, 2024. do not change this unless explicitly requested by the user
        messages,
        ...(useJson && { response_format: { type: "json_object" as const } }),
        temperature: 0.3,
        max_tokens: 2000,
      });

      const content = response.choices[0].message.content;
      return useJson ? JSON.parse(content || "{}") : content;
    } catch (error) {
      console.error(`Error in ${this.agentType} agent:`, error);
      throw new Error(`Failed to process query with ${this.agentType} agent`);
    }
  }

  protected calculateConfidence(response: any): number {
    if (response.confidence) return Math.min(100, Math.max(0, response.confidence));
    
    const hasResponse = response.response && response.response.length > 10;
    const hasSources = response.sources && response.sources.length > 0;
    
    if (hasResponse && hasSources) return 85;
    if (hasResponse) return 70;
    return 50;
  }
}
